README.md
# Clocker


[![MoeKit](https://moekit.timo.today/badge/clocker)](https://moekit.timo.today/package/clocker)  [![Build Status](https://travis-ci.org/MoeKit/clocker.svg?branch=master)](https://travis-ci.org/MoeKit/clocker)  [![Coverage Status](https://coveralls.io/repos/MoeKit/clocker/badge.svg)](https://coveralls.io/r/MoeKit/clocker)
---


## 使用

构造器的时间参数可以为

+ Date 对象
+ 毫秒数
+ 格式化的时间字符串如 `YYYY/MM/DD` ,`MM/DD/YYYY`,`YYYY/MM/DD hh:mm:ss`,`MM/DD/YYYY hh:mm:ss`

```javascript
var Clocker = require('clocker');
var ck = new Clock('2016-06-01',{
    precision:100 // 默认为100ms
}).start();
ck.on('tick',function(event){
	// do what you want
});
```

## API

### 函数

+ `getOffset` 获取时间差，一般用于不需要实时计数时，此时不需要调用start方法

```
var Clocker = require('clocker');
var offset = new Clock('2016-06-01').getOffset();

```

+ `pause` 暂停

+ `start` 开始

+ `stop` 结束

+ `resume` 重新开始

### 事件

+ `tick` 定时触发，按照时间精度的设置间隔触发,默认interval时间间隔为`100ms`

+ `finish` 计时结束

+ `stop` 停止


事件的参数`event`说明：

```json
{
    type:           '{String} The namespaced event type {update,finish,stop}.countdown',
    strftime:       '{Function} The formatter function',
    finalDate:      '{Date} The parsed final date native object',
    offset: {
        seconds:    '{int} Seconds left for the next minute',
        minutes:    '{int} Minutes left for the next hour',
        hours:      '{int} Hours left until next day',
        days:       '{int} Days left until next week',
        totalDays:  '{int} Total amount of days left until final date',
        weeks:      '{int} Weeks left until the final date',
        months:     '{int} Months left until final date' ,
        years:      '{int} Year left until final date'
    }
}
```

通过event的offset可以获取当前的时间差,通过strftime函数可以直接格式化文本。

如

```javascript
var Clocker = require('index');
var ck = new Clock('2016-06-01');
ck.on('tick',function(event){
	var html = event.strftime('%W weeks %D days %H h %M min %S sec');
    // => 1 week 2days 3h 04 min 05 sec
});
```

## 格式化Directives说明
| 字符 | 说明   |
|---|---|
|%Y|	Years left *|
|%m	|	Monts left * |
|%w |	Weeks left |
|%d |	Days left (taking away weeks) |
|%D |	Total amount of days left |
|%H	|	Hours left |
|%M	|	Minutes left |
|%S	|	Seconds left |



## 感谢

该模块Fork自:http://hilios.github.io/jQuery.countdown/ , 但去除了`jQuery`依赖,同时用`Eventor`修改了事件机制。

HISTORY.md
# History

---

## 2.0.4

`tag:new` 必须手动调用`start`方法，以减少只使用`getOffset`时的无必要定时器

`tag:new` 可以指定`precision`参数，当不需要精确到秒时，比如只需要到分钟，那么可以指定为1分钟

## 2.0.3

`tag:improved` 增加`getOffset`方法，用于不需要tick的场景

## 2.0.2

`tag:new` 初次提交



tests/countdown-spec.js
var Clocker = require('../index');
describe('countdown', function() {
    it('normal usage', function() {
        var clocker = new Clocker('2016/06/01').on('tick',function(event){
           console.log('tick'); 
        });
    });
});

index.js
var instances = [],
    matchers = [];
// Miliseconds
matchers.push(/^[0-9]*$/.source);
// Month/Day/Year [hours:minutes:seconds]
matchers.push(/([0-9]{1,2}\/){2}[0-9]{4}( [0-9]{1,2}(:[0-9]{2}){2})?/
    .source);
// Year/Day/Month [hours:minutes:seconds] and
// Year-Day-Month [hours:minutes:seconds]
matchers.push(/[0-9]{4}([\/\-][0-9]{1,2}){2}( [0-9]{1,2}(:[0-9]{2}){2})?/
    .source);
// Cast the matchers to a regular expression object
matchers = new RegExp(matchers.join('|'));
// Parse a Date formatted has String to a native object

function parseDateString(dateString) {
    // Pass through when a native object is sent
    if (dateString instanceof Date) {
        return dateString;
    }
    // Caste string to date object
    if (String(dateString).match(matchers)) {
        // If looks like a milisecond value cast to number before
        // final casting (Thanks to @msigley)
        if (String(dateString).match(/^[0-9]*$/)) {
            dateString = Number(dateString);
        }
        // Replace dashes to slashes
        if (String(dateString).match(/\-/)) {
            dateString = String(dateString).replace(/\-/g, '/');
        }
        return new Date(dateString);
    } else {
        throw new Error('Couldn\'t cast `' + dateString +
            '` to a date object.');
    }
}
// Map to convert from a directive to offset object property
var DIRECTIVE_KEY_MAP = {
    'Y': 'years',
    'm': 'months',
    'w': 'weeks',
    'D': 'days',
    'H': 'hours',
    'M': 'minutes',
    'S': 'seconds',
};
// Returns an escaped regexp from the string

function escapedRegExp(str) {
    var sanitize = str.toString().replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');
    return new RegExp(sanitize);
}
// Time string formatter

function strftime(offsetObject) {
    return function(format) {
        var directives = format.match(/%(-|!)?[A-Z]{1}(:[^;]+;)?/gi);
        if (directives) {
            for (var i = 0, len = directives.length; i < len; ++i) {
                var directive = directives[i]
                    .match(/%(-|!)?([a-zA-Z]{1})(:[^;]+;)?/),
                    regexp = escapedRegExp(directive[0]),
                    modifier = directive[1] || '',
                    plural = directive[3] || '',
                    value = null;
                // Get the key
                directive = directive[2];
                // Swap shot-versions directives
                if (DIRECTIVE_KEY_MAP.hasOwnProperty(directive)) {
                    value = DIRECTIVE_KEY_MAP[directive];
                    value = Number(offsetObject[value]);
                }
                if (value !== null) {
                    // Pluralize
                    if (modifier === '!') {
                        value = pluralize(plural, value);
                    }
                    // Add zero-padding
                    if (modifier === '') {
                        if (value < 10) {
                            value = '0' + value.toString();
                        }
                    }
                    // Replace the directive
                    format = format.replace(regexp, value.toString());
                }
            }
        }
        format = format.replace('%_M1',offsetObject.minutes_1)
                        .replace('%_M2',offsetObject.minutes_2)
                        .replace('%_S1',offsetObject.seconds_1)
                        .replace('%_S2',offsetObject.seconds_2)
                        .replace('%_H1',offsetObject.hours_1)
                        .replace('%_H2',offsetObject.hours_2)
                        .replace('%_D1',offsetObject.days_1)
                        .replace('%_D2',offsetObject.days_2);
        format = format.replace(/%%/, '%');
        return format;
    };
}
// Pluralize

function pluralize(format, count) {
    var plural = 's',
        singular = '';
    if (format) {
        format = format.replace(/(:|;|\s)/gi, '').split(/\,/);
        if (format.length === 1) {
            plural = format[0];
        } else {
            singular = format[0];
            plural = format[1];
        }
    }
    if (Math.abs(count) === 1) {
        return singular;
    } else {
        return plural;
    }
}

function splitNumber (number){
    number = number+'';
    number = (number.length===1?('0'+number):number)+'';
    return number.split('');
}

// The Final Countdown
var Countdown = function(finalDate, option) {
    option = option || {};
    this.PRECISION = option.precision || 100; // 0.1 seconds, used to update the DOM
    this.interval = null;
    this.offset = {};
    // Register this instance
    this.instanceNumber = instances.length;
    instances.push(this);
    // Set the final date and start
    this.setFinalDate(finalDate);
};
var Eventor = require('eventor');
Eventor.mixTo(Countdown);
var pro = Countdown.prototype;

var fns = {
    start: function() {
        if (this.interval !== null) {
            clearInterval(this.interval);
        }
        var self = this;
        this.update();
        this.interval = setInterval(function() {
            self.update.call(self);
        }, this.PRECISION);
        return this;
    },
    stop: function() {
        clearInterval(this.interval);
        this.interval = null;
        this._dispatchEvent('stoped');
        return this;
    },
    toggle: function() {
        if (this.interval) {
            this.stop();
        } else {
            this.start();
        }
        return this;
    },
    pause: function() {
        return this.stop();
    },
    resume: function() {
        return this.start();
    },
    remove: function() {
        this.stop.call(this);
        instances[this.instanceNumber] = null;
    },
    setFinalDate: function(value) {
        this.finalDate = parseDateString(value); // Cast the given date
        return this;
    },
    getOffset: function(){
        this.totalSecsLeft = this.finalDate.getTime() - new Date().getTime(); // In miliseconds
        this.totalSecsLeft = Math.ceil(this.totalSecsLeft / 1000);
        this.totalSecsLeft = this.totalSecsLeft < 0 ? 0 : this.totalSecsLeft;
        // Calculate the offsets
        return {
            seconds: this.totalSecsLeft % 60,
            minutes: Math.floor(this.totalSecsLeft / 60) % 60,
            hours: Math.floor(this.totalSecsLeft / 60 / 60) % 24,
            days: Math.floor(this.totalSecsLeft / 60 / 60 / 24),
            weeks: Math.floor(this.totalSecsLeft / 60 / 60 / 24 / 7),
            months: Math.floor(this.totalSecsLeft / 60 / 60 / 24 / 30),
            years: Math.floor(this.totalSecsLeft / 60 / 60 / 24 / 365)
        };
    },
    update: function() {
        // Calculate the offsets
        this.offset = this.getOffset();
        // split offset only for days, hours, minutes, seconds and two number like 45, do not support 100

        var list = ['days','hours','minutes','seconds'];

        for(var i=0;i<list.length;i++){
            var key = list[i];
            var numbers = splitNumber(this.offset[key]);
            this.offset[key+'_1'] = numbers[0];
            this.offset[key+'_2'] = numbers[1];
        }
        // Dispatch an event
        if (this.totalSecsLeft === 0) {
            this.stop();
            this._dispatchEvent('finish');
        } else {
            this._dispatchEvent('update');
        }
        return this;
    },
    _dispatchEvent: function(eventName) {
        var event = {};
        event.finalDate = this.finalDate;
        event.offset = this.offset;
        event.strftime = strftime(this.offset);
        this.emit(eventName, event);
        this.emit('tick',event);
    }
}

for (var i in fns) {
    pro[i] = fns[i];
}

module.exports = Countdown;

src/countdown.js
var $ = window.jQuery || require('zepto');
console.log($);
module.exports = $;
/*!
 * The Final Countdown for jQuery v2.0.2 (http://hilios.github.io/jQuery.countdown/)
 * Copyright (c) 2013 Edson Hilios
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

var PRECISION = 100;
var instances = [],
    matchers = [];
matchers.push(/^[0-9]*$/.source);
matchers.push(/([0-9]{1,2}\/){2}[0-9]{4}( [0-9]{1,2}(:[0-9]{2}){2})?/.source);
matchers.push(/[0-9]{4}(\/[0-9]{1,2}){2}( [0-9]{1,2}(:[0-9]{2}){2})?/.source);
matchers = new RegExp(matchers.join("|"));

function parseDateString(dateString) {
    if (dateString instanceof Date) {
        return dateString;
    }
    if (String(dateString).match(matchers)) {
        if (String(dateString).match(/^[0-9]*$/)) {
            dateString = Number(dateString);
        }
        return new Date(dateString);
    } else {
        throw new Error("Couldn't cast `" + dateString + "` to a date object.");
    }
}
var DIRECTIVE_KEY_MAP = {
    Y: "years",
    m: "months",
    w: "weeks",
    d: "days",
    D: "totalDays",
    H: "hours",
    M: "minutes",
    S: "seconds"
};

function strftime(offsetObject) {
    return function(format) {
        var directives = format.match(/%(-|!)?[A-Z]{1}(:[^;]+;)?/gi);
        if (directives) {
            for (var i = 0, len = directives.length; i < len; ++i) {
                var directive = directives[i].match(/%(-|!)?([a-zA-Z]{1})(:[^;]+;)?/),
                    regexp = new RegExp(directive[0]),
                    modifier = directive[1] || "",
                    plural = directive[3] || "",
                    value = null;
                directive = directive[2];
                if (DIRECTIVE_KEY_MAP.hasOwnProperty(directive)) {
                    value = DIRECTIVE_KEY_MAP[directive];
                    value = Number(offsetObject[value]);
                }
                if (value !== null) {
                    if (modifier === "!") {
                        value = pluralize(plural, value);
                    }
                    if (modifier === "") {
                        if (value < 10) {
                            value = "0" + value.toString();
                        }
                    }
                    format = format.replace(regexp, value.toString());
                }
            }
        }
        format = format.replace(/%%/, "%");
        return format;
    };
}

function pluralize(format, count) {
    var plural = "s",
        singular = "";
    if (format) {
        format = format.replace(/(:|;|\s)/gi, "").split(/\,/);
        if (format.length === 1) {
            plural = format[0];
        } else {
            singular = format[0];
            plural = format[1];
        }
    }
    if (Math.abs(count) === 1) {
        return singular;
    } else {
        return plural;
    }
}
var Countdown = function(el, finalDate, callback) {
    this.el = el;
    this.$el = $(el);
    this.interval = null;
    this.offset = {};
    this.setFinalDate(finalDate);
    this.instanceNumber = instances.length;
    instances.push(this);
    this.$el.data("countdown-instance", this.instanceNumber);
    if (callback) {
        this.$el.on("update.countdown", callback);
        this.$el.on("stoped.countdown", callback);
        this.$el.on("finish.countdown", callback);
    }
    this.start();
};
$.extend(Countdown.prototype, {
    start: function() {
        if (this.interval !== null) {
            throw new Error("Countdown is already running!");
        }
        var self = this;
        this.update();
        this.interval = setInterval(function() {
            self.update.call(self);
        }, PRECISION);
    },
    stop: function() {
        clearInterval(this.interval);
        this.interval = null;
        this.dispatchEvent("stoped");
    },
    pause: function() {
        this.stop.call(this);
    },
    resume: function() {
        this.start.call(this);
    },
    remove: function() {
        this.stop();
        delete instances[this.instanceNumber];
    },
    setFinalDate: function(value) {
        this.finalDate = parseDateString(value);
    },
    update: function() {
        if (this.$el.closest("html").length === 0) {
            this.remove();
            return;
        }
        this.totalSecsLeft = this.finalDate.valueOf() - new Date().valueOf();
        this.totalSecsLeft = Math.ceil(this.totalSecsLeft / 1e3);
        this.totalSecsLeft = this.totalSecsLeft < 0 ? 0 : this.totalSecsLeft;
        this.offset = {
            seconds: this.totalSecsLeft % 60,
            minutes: Math.floor(this.totalSecsLeft / 60) % 60,
            hours: Math.floor(this.totalSecsLeft / 60 / 60) % 24,
            days: Math.floor(this.totalSecsLeft / 60 / 60 / 24) % 7,
            totalDays: Math.floor(this.totalSecsLeft / 60 / 60 / 24),
            weeks: Math.floor(this.totalSecsLeft / 60 / 60 / 24 / 7),
            months: Math.floor(this.totalSecsLeft / 60 / 60 / 24 / 30),
            years: Math.floor(this.totalSecsLeft / 60 / 60 / 24 / 365)
        };
        if (this.totalSecsLeft === 0) {
            this.stop();
            this.dispatchEvent("finish");
        } else {
            this.dispatchEvent("update");
        }
    },
    dispatchEvent: function(eventName) {
        var event = $.Event(eventName + ".countdown");
        event.finalDate = this.finalDate;
        event.offset = $.extend({}, this.offset);
        event.strftime = strftime(this.offset);
        this.$el.trigger(event);
    }
});
$.fn.countdown = function() {
    var argumentsArray = Array.prototype.slice.call(arguments, 0);
    return this.each(function() {
        var instanceNumber = $(this).data("countdown-instance");
        console.log(instanceNumber)
        if (instanceNumber !== undefined) {
            var instance = instances[instanceNumber],
                method = argumentsArray[0];
            if (Countdown.prototype.hasOwnProperty(method)) {
                instance[method].apply(instance, argumentsArray.slice(1));
            } else if (String(method).match(/^[$A-Z_][0-9A-Z_$]*$/i) === null) {
                instance.setFinalDate.call(instance, method);
            } else {
                $.error("Method %s does not exist on jQuery.countdown".replace(/\%s/gi, method));
            }
        } else {
            new Countdown(this, argumentsArray[0], argumentsArray[1]);
        }
    });
};

console.log($.fn.countdown)